usersession/userd/launcher.go

```diff
@@ -123,6 +131,105 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
 	return nil
 }
 
+// OpenDesktopEntry implements the 'OpenDesktopEntry' method of the 'io.snapcraft.Launcher'
+// DBus interface. Before the provided desktop_file is parsed it is validated against a list
+// of allowed locations.
+func (s *Launcher) OpenDesktopEntry(desktop_file_id string, sender dbus.Sender) *dbus.Error {
+
+	return s.OpenDesktopEntryEnv(desktop_file_id, []string{}, sender)
+}
+
+// OpenDesktopEntryEnv implements the 'OpenDesktopEntryEnv' method of the 'io.snapcraft.Launcher'
+// DBus interface.
+func (s *Launcher) OpenDesktopEntryEnv(desktop_file_id string, env []string, sender dbus.Sender) *dbus.Error {
+	exec_command, err := s.readExecCommandFromDesktopFile(s.desktopFileIdToFilename(desktop_file_id))
+
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	// Passing exec variables between confined snaps raises unanswered questions and they are not required
+	// for the simple cases.  For now, we don't have support for passing them in the dbus API and truncate
+	// the exec command at the first exec variable.
+	// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#exec-variables
+	exec_command = strings.SplitN(exec_command, "%", 2)[0]
+
+	args := strings.Split(exec_command, " ")
```

Comment:
This does not handle `Exec` lines that make use of shell quoting.  From the look of it, this library could help:

https://godoc.org/github.com/google/shlex

Reply:
OK, thanks

Reply:
@jhenstridge I'm missing something, it works locally, but CI says:
```
src/github.com/snapcore/snapd/usersession/userd/launcher.go:35:2: cannot find package "github.com/google/shlex" in any of:
	/root/rpmbuild/BUILD/snapd-1337.2.42~pre2/src/github.com/snapcore/snapd/vendor/github.com/google/shlex (vendor tree)
	/usr/lib/golang/src/github.com/google/shlex (from $GOROOT)
	/root/rpmbuild/BUILD/snapd-1337.2.42~pre2/src/github.com/google/shlex (from $GOPATH)
	/usr/share/gocode/src/github.com/google/shlex
error: Bad exit status from /var/tmp/rpm-tmp.BByDfB (%build)
```
Any suggestions?

Reply:
The package would need to be vendored, adding to `vendor/vendor.json`.  I think the `govendor fetch` command can do this for you.

Reply:
Like this? https://github.com/snapcore/snapd/pull/7490/files#diff-bd290170e2912d3e8694db1a151066e5L27